api/webhook/validate_auth.go

```diff
@@ -0,0 +1,199 @@
+// limitations under the License.
+
+package webhook
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"reflect"
+
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	authv1 "k8s.io/api/authorization/v1"
+	authorizationv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/pkg/selector"
+)
+
+var authLog = ctrl.Log.WithName("validate-auth")
+
+// +kubebuilder:webhook:path=/validate-auth,mutating=false,failurePolicy=fail,groups=chaos-mesh.org,resources=*,verbs=create;update,versions=v1alpha1,name=vauth.kb.io
+
+// AuthValidator validates the authority
+type AuthValidator struct {
+	Enable  bool
+	Client  client.Client
+	Reader  client.Reader
+	AuthCli *authorizationv1.AuthorizationV1Client
+
+	decoder *admission.Decoder
+
+	ClusterScoped     bool
+	TargetNamespace   string
+	AllowedNamespaces string
+	IgnoredNamespaces string
+}
+
+// AuthValidator admits a pod iff a specific annotation exists.
+func (v *AuthValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if !v.Enable {
+		return admission.Allowed("")
+	}
+
+	needAuth := true
+	username := req.UserInfo.Username
+	groups := req.UserInfo.Groups
+	chaosKind := req.Kind.Kind
+
+	var chaos, oldChaos v1alpha1.ChaosValidator
+
+	switch chaosKind {
+	case v1alpha1.KindPodChaos:
+		chaos = &v1alpha1.PodChaos{}
+		oldChaos = &v1alpha1.PodChaos{}
+
+	case v1alpha1.KindIoChaos:
+		chaos = &v1alpha1.IoChaos{}
+		oldChaos = &v1alpha1.IoChaos{}
+
+	case v1alpha1.KindNetworkChaos:
+		chaos = &v1alpha1.NetworkChaos{}
+		oldChaos = &v1alpha1.NetworkChaos{}
+
+	case v1alpha1.KindTimeChaos:
+		chaos = &v1alpha1.TimeChaos{}
+		oldChaos = &v1alpha1.TimeChaos{}
+
+	case v1alpha1.KindKernelChaos:
+		chaos = &v1alpha1.KernelChaos{}
+		oldChaos = &v1alpha1.KernelChaos{}
+
+	case v1alpha1.KindStressChaos:
+		chaos = &v1alpha1.StressChaos{}
+		oldChaos = &v1alpha1.StressChaos{}
+
+	case v1alpha1.KindDNSChaos:
+		chaos = &v1alpha1.DNSChaos{}
+		oldChaos = &v1alpha1.DNSChaos{}
+
+	case v1alpha1.KindAwsChaos, v1alpha1.KindPodNetworkChaos:
+		needAuth = false
+	default:
+		err := fmt.Errorf("kind %s is not support", chaosKind)
+
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	if !needAuth {
+		return admission.Allowed("")
+	}
+
+	err := v.decoder.Decode(req, chaos)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	specs := chaos.GetSelectSpec()
+
+	if req.Operation == admissionv1beta1.Update {
+		// when selector is not changed, don't need to validate auth again
+
+		err = v.decoder.DecodeRaw(req.OldObject, oldChaos)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		oldSpecs := oldChaos.GetSelectSpec()
+		selectSpecChanged := false
+		if len(specs) == len(oldSpecs) {
```

Comment:
So if `len(specs) != len(oldSpecs)`, `selectSpecChanged` would still be `false`?

Reply:
fixed in 01c9404